server/data/hmppsAuditClient.ts

```diff
@@ -0,0 +1,70 @@
+import { SendMessageCommand, SQSClient } from '@aws-sdk/client-sqs'
+import logger from '../../logger'
+
+export interface AuditEvent {
+  what: string
+  who: string
+  subjectId?: string
+  subjectType?: string
+  correlationId?: string
+  details?: object
+}
+
+export interface SqsMessage {
+  what: string
+  who: string
+  when: string
+  service: string
+  subjectId?: string
+  subjectType?: string
+  correlationId?: string
+  details?: object
+}
+
+export interface AuditClientConfig {
+  queueUrl: string
+  region: string
+  serviceName: string
+  enabled: boolean
+}
+
+export default class HmppsAuditClient {
+  private sqsClient: SQSClient
+
+  private queueUrl: string
+
+  private serviceName: string
+
+  private enabled: boolean
+
+  constructor(config: AuditClientConfig) {
+    this.enabled = config.enabled
+    this.queueUrl = config.queueUrl
+    this.serviceName = config.serviceName
+    this.sqsClient = new SQSClient({ region: config.region })
+  }
+
+  async sendMessage(event: AuditEvent, throwOnError: boolean = false) {
+    if (!this.enabled) return null
+
+    const sqsMessage: SqsMessage = {
+      ...event,
+      service: this.serviceName,
+      when: new Date().toISOString(),
+    }
+
+    try {
+      const messageResponse = await this.sqsClient.send(
+        new SendMessageCommand({ MessageBody: JSON.stringify(sqsMessage), QueueUrl: this.queueUrl }),
+      )
+
+      logger.info(`HMPPS Audit SQS message sent (${messageResponse.MessageId})`)
+
+      return messageResponse
+    } catch (error) {
+      logger.error('Error sending HMPPS Audit SQS message, ', error)
+      if (throwOnError) throw error
```

Comment:
A question was raised last week about what we should do in the unlikely event an audit event fails. Martin Ballhatchet suggested it would be best to show an error since it risks user actions going unaudited.

The reason for this parameter is so teams can decide whether audit errors should be thrown or simply logged.

An alternative approach here might be for the client to always throw an error and for teams to decide how they want to handle that exception in the service layer?

Reply:
I think this is nice - defaulting to throw but allowing overriding.